app/forms/reactivate_profile_form.rb

```diff
@@ -0,0 +1,69 @@
+class ReactivateProfileForm
+  include ActiveModel::Model
+
+  validates :recovery_code, :password, presence: true
+  validate :validate_password_reset_profile
+  validate :validate_password
+  validate :validate_recovery_code
+
+  attr_accessor :recovery_code, :password
+  attr_reader :user
+
+  def initialize(user, attrs = {})
+    @user = user
+    super attrs
+  end
+
+  def submit(flash)
+    if valid?
+      flash[:recovery_code] = reencrypt_pii
+      true
+    else
+      clear_fields
+      false
+    end
+  end
+
+  protected
+
+  def reencrypt_pii
+    profile = user.password_reset_profile
+    pii = profile.recover_pii(recovery_code)
+    user_access_key = user.unlock_user_access_key(password)
+    recovery_code = profile.encrypt_pii(user_access_key, pii)
+    profile.deactivation_reason = :not_deactivated
+    profile.save!
+    recovery_code
```

Comment:
there is a lot happening in this method! Thoughts on having a class that just takes care of this logic / can be testing in isolation?